pip/pip-367.md

```diff
@@ -0,0 +1,193 @@
+	// directly to the DLQ in accordance with the configured [DLQPolicy]. If there is no DLQPolicy configured on the
+	// consumer, this method will error.
+	Term(Message) error
+}
+```
+
+```java
+public interface Consumer<T> {
+
+    /**
+     * Non-retryable failure to process a single message.
+     * <p>When a message is "terminally negatively acked" it will not be marked for redelivery. Instead,
+     * it will be routed directly to the DLQ in accordance with the configured {@link DeadLetterPolicy}
+     *
+     * @param messageId {@link MessageId} to be individual acknowledged
+     * @throws PulsarClientException.InvalidConfigurationException
+     *             if the consumer did not set the {@link DeadLetterPolicy}
+     */
+    void terminate(MessageId messageId) throws PulsarClientException;
+
+    void terminate(Message<?> message) throws PulsarClientException;
+}
+
+```
+
+Here is the existing flow that a Consumer undergoes in order to route a message to the DLQ.
+```mermaid
+sequenceDiagram
+    participant Broker
+    participant Consumer
+    loop
+        Broker->>Consumer: Receive message A
+        alt attempts > max?
+            Consumer->>Broker: ACK
+            Consumer->>Broker: Produce message A to DLQ
+            Note right of Broker: break
+        else
+            Consumer->>Consumer: Process message A
+            Note right of Consumer: Processing message A fails
+            Consumer->>Consumer: NACK
+            Consumer->>Broker: Resend message A
+            Broker->>Broker: attempts++
+        end
+    end
+```
+
+Here is the same flow augmented to allow a Consumer to terminally fail a message and route it to the DLQ without having
+to retry up to several times.
+```mermaid
+sequenceDiagram
+    participant Broker
+    participant Consumer
+    loop
+        Broker->>Consumer: Receive message A
+        alt attempts > max?
+            Consumer->>Broker: ACK
+            Consumer->>Broker: Produce message A to DLQ
+            Note right of Broker: break
+        else
+            Consumer->>Consumer: Process message A
+            Note right of Consumer: Processing message A fails
+            alt Terminal failure?
+                Consumer->>Consumer: TERM
+                Consumer->>Broker: ACK
+                Consumer->>Broker: Produce message A to DLQ
+                Note right of Broker: break
+            else
+                Consumer->>Consumer: NACK
+                Consumer->>Broker: Resend message A
+                Broker->>Broker: attempts++
+            end
+        end
+    end
+```
+
+## Public-facing Changes
+
+### Public API
+
+N/A
+
+### Binary protocol
+
+N/A
+
+### Configuration
+
+N/A
+
+### CLI
+
+N/A
+
+### Metrics
+
+If the client would like to track the number of calls to the `Term` method, we could add a new metric
+`pulsar_client_consumer_terms`. This metric should in theory track exactly with the
```

Comment:
`pulsar_client_consumer_message_terms`
It is recommended to keep the naming conventions consistent.
https://github.com/apache/pulsar/blob/d4bbf10f58771e2d43e576dc3422e502834b1de4/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L429-L430

Reply:
They are all defined with underscores in the Go client library.

Reply:
The Java client is currently using OpenTelemetry's naming conventions. I suggest aligning it with the existing naming style.

Reply:
I think that moving the Go client to be OTel compliant is outside the scope of this change. I think I would rather the Go client be consistent with itself. Changing all of the metrics to be OTel compliant can be done in a single non-backwards compatible change.

Reply:
Yes, I meant that the Java client should follow the current naming conventions. It would be best to also add a description for Java metrics.